p2p/src/network/pubsub/p2p_network_pubsub_reducer.rs

```diff
@@ -123,14 +123,26 @@ impl P2pNetworkPubsubState {
                 peer_id,
                 data,
                 seen_limit,
+                addr,
                 ..
             } => {
                 pubsub_state.reduce_incoming_data(&peer_id, data, meta.time())?;
 
-                let dispatcher: &mut Dispatcher<Action, State> = state_context.into_dispatcher();
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                let p2p_state: &P2pState = state.substate()?;
+                let state = &p2p_state.network.scheduler.broadcast_state;
+                let Some(state) = state.clients.get(&peer_id) else {
+                    // TODO: investigate, cannot reproduce this
+                    // bug_condition!("{:?} not found in state.clients", peer_id);
+                    return Ok(());
+                };
+
+                let messages = state.incoming_messages.clone();
```

Comment:
I am worried about all this cloning here and bellow, is it really needed? cannot the messages use `Arc` ?

probably out of scope for this requests, but worth exploring given that we are already having memory issues because of data (mis?)handling in the p2p layer

Reply:
If the answer is not immediately obvious please create an issue for that task so that you don't forget later.

Reply:
This was already cloned in original but in effectful effects, so it won't make difference, but most likely it could be optimized to reduce memory